common/state-machine-lib/src/test/java/org/eclipse/dataspaceconnector/common/statemachine/StateMachineTest.java

```diff
@@ -0,0 +1,125 @@
+
+class StateMachineTest {
+
+    private final WaitStrategy waitStrategy = mock(WaitStrategy.class);
+    private final Monitor monitor = mock(Monitor.class);
+
+    @BeforeEach
+    void setUp() {
+        when(waitStrategy.waitForMillis()).thenReturn(1L);
+    }
+
+    @Test
+    void shouldExecuteProcessorsAsyncAndCanBeStopped() throws InterruptedException {
+        var latch = new CountDownLatch(2);
+        var processor = mock(EntitiesProcessor.class);
+        when(processor.run()).thenAnswer(i -> {
+            latch.countDown();
+            Thread.sleep(100L);
+            return 1L;
+        });
+        var loop = StateMachine.Builder.newInstance("test", monitor, waitStrategy)
+                .processor(processor)
+                .shutdownTimeout(1)
+                .build();
+
+        loop.start();
+        assertThat(latch.await(1, SECONDS)).isTrue();
+        verify(processor, atLeastOnce()).run();
+
+        assertThat(loop.stop()).succeedsWithin(2, SECONDS);
+        verifyNoMoreInteractions(processor);
+    }
+
+    @Test
+    void shouldNotWaitForSomeTimeIfTheresAtLeastOneProcessedEntity() throws InterruptedException {
+        var processor = mock(EntitiesProcessor.class);
+        when(processor.run()).thenReturn(1L);
+        var latch = new CountDownLatch(1);
+        doAnswer(i -> {
+            latch.countDown();
+            return 1L;
+        }).when(waitStrategy).success();
+        var loop = StateMachine.Builder.newInstance("test", monitor, waitStrategy)
+                .processor(processor)
+                .build();
+
+        loop.start();
+
+        assertThat(latch.await(1, SECONDS)).isTrue();
+        verify(waitStrategy, never()).waitForMillis();
+        verify(waitStrategy, atLeastOnce()).success();
+    }
+
+    @Test
+    void shouldWaitForSomeTimeIfNoEntityIsProcessed() throws InterruptedException {
+        var processor = mock(EntitiesProcessor.class);
+        when(processor.run()).thenReturn(0L);
+        var latch = new CountDownLatch(1);
+        var waitStrategy = mock(WaitStrategy.class);
+        doAnswer(i -> {
+            latch.countDown();
+            return 0L;
+        }).when(waitStrategy).success();
+        var loop = StateMachine.Builder.newInstance("test", monitor, waitStrategy)
+                .processor(processor)
+                .build();
+
+        loop.start();
+
+        assertThat(latch.await(1, SECONDS)).isTrue();
+        verify(waitStrategy, atLeastOnce()).waitForMillis();
+        verify(waitStrategy, atLeastOnce()).success();
+    }
+
+    @Test
+    void shouldExitWithAnExceptionIfProcessorExitsWithAnUnrecoverableError() {
+        var processor = mock(EntitiesProcessor.class);
+        when(processor.run()).thenThrow(new Error("unrecoverable"));
+        var loop = StateMachine.Builder.newInstance("test", monitor, waitStrategy)
+                .processor(processor)
+                .build();
+
+        assertThat(loop.start()).succeedsWithin(1, SECONDS);
```

Comment:
With parallel builds and Jacoco we might have sporadic slow tests. Is there a strong reason for such short timeouts?

Reply:
I keep them shorts because I felt like it's a fast operation and it shouldn't take more than a second to complete also in a slow test environment. Let's see how it will behave.

Reply:
and, another thing that could be improved, I had to modify the monitor implementation to have the output, I think we should have the logs enabled by default and maybe disable them on the CI